[MRG+1] Add a new class RegressorChain similar to ClassifierChain

thechargedneutron opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

#9246 
<!-- Example: Fixes #1234 -->


#### What does this implement/fix? Explain your changes.


The changes made are as follows
1. Removed local variable class_
2. Removed decision_function
3. Removed predict_proba

#### Any other comments?

Suggestions are welcomed.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thechargedneutron committed:
Update multioutput.py

thechargedneutron committed:
Update multioutput.py

TomDLT commented:
Would it be possible to avoid redundant code between `ClassifierChain` and `RegressorChain`?
Something like:
```py
class BaseChain(BaseEstimator)
    def __init__(self, base_estimator, order=None, cv=None, random_state=None):
        ...
    def fit(self, X, Y):
        ...
    def predict(self, X):
        ...

class ClassifierChain(BaseChain):
    """docstring"""
    @if_delegate_has_method('base_estimator')
    def predict_proba(self, X):
        ...
    @if_delegate_has_method('base_estimator')
    def decision_function(self, X):
        ...

class RegressorChain(BaseChain):
    """docstring"""
    ...
```

You will also need to fix the tests, maybe add new tests, add a docstring in `RegressorChain`, add an entry in the documentation, and add `RegressorChain` in `doc/modules/classes.rst`.

Basically, look for every appearance of `ClassifierChain` (with for example `git grep ClassifierChain`), and add something similar with `RegressorChain` _if it's relevant_.

Your code is also not PEP8 compliant.

thechargedneutron committed:
Progress Updated

* Update classes.rst

* Update multiclass.rst

* Update multioutput.py

* Update testing.py

thechargedneutron commented:
Work Progress: 
1. Edited multioutput.py to make it compatible with PEP8. (Number of characters > 79 ignored)
2. Added the function in docs. (Except in whats_new. Do I have to add there also??)
3. Added "docstring" in RegressorChain.

Need help in adding more tests. 
Could "generate_multilabel_dataset_with_correlations()" be used to generate tests for Regression, I doubt.

jnothman commented:
I don't think we're going to merge this for the upcoming release. As we
want to focus on the release, please ping us in a couple of weeks

On 1 July 2017 at 06:57, Kumar Ashutosh <notifications@github.com> wrote:

> Work Progress:
>
>    1. Edited multioutput.py to make it compatible with PEP8. (Number of
>    characters > 79 ignored)
>    2. Added the function in docs. (Except in whats_new. Do I have to add
>    there also??)
>    3. Added "docstring" in RegressorChain.
>
> Need help in adding more tests.
> Could "generate_multilabel_dataset_with_correlations()" be used to
> generate tests for Regression, I doubt.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-312371161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68_FG-zmrlavuzQmdyCw2PELKcB8ks5sJWEdgaJpZM4OKxgm>
> .
>

jmschrei commented:
I would also support create a BaseChain class that is common to both regressors and classifiers. It seems to fit with the other models we have.

jnothman commented:
Are you still working on this @thechargedneutron

thechargedneutron commented:
Yes sir, since you were focussing on new release, I did not commit. I wish to continue the work. I'll update the PR in a day.

thechargedneutron committed:
Adds Common BaseClass for RegressorChain and ClassifierChain

thechargedneutron commented:
The common BaseChain class has been added. I tried resolving the conflicting files but somehow did not work. Kindly review the changes and remove the conflicts. 
Also I need suggestions on how to add tests.
Thanks

jnothman commented:
The tests should look very similar to those for ClassifierChain, or you
might find cases where one test function can loop over both the classifier
and regressor case.

On 31 Jul 2017 3:39 pm, "Kumar Ashutosh" <notifications@github.com> wrote:

> The common BaseChain class has been added. I tried resolving the
> conflicting files but somehow did not work. Kindly review the changes and
> remove the conflicts.
> Also I need suggestions on how to add tests.
> Thanks
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-318972897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz680nU23pN1i_LcK2RjEU9Nb-8XNCks5sTWingaJpZM4OKxgm>
> .
>

jnothman committed:
Merge branch 'master' into master

thechargedneutron committed:
Tests added

thechargedneutron committed:
Merge branch 'master' of https://github.com/thechargedneutron/scikit-learn

thechargedneutron commented:
The tests have been added. SVR was not added in the RegressorChain() tests as the function decision_function() is deprecated and will be removed in 0.19.  If there's a mistake or something more is to be added, kindly let me know.
Thanks.

jnothman commented:
yes, decision_function is not used in regressors

On 2 Aug 2017 5:23 am, "Kumar Ashutosh" <notifications@github.com> wrote:

> The tests have been added. SVR was not added in the RegressorChain() tests
> as the function decision_function() is deprecated and will be removed in
> 0.19. If there's a mistake or something more is to be added, kindly let me
> know.
> Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-319470839>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-8ucPWmG92V_wL3LI1-lFccFl5aks5sT3tKgaJpZM4OKxgm>
> .
>

thechargedneutron commented:
Is there anything else to be worked upon? And why are the CI build failing?
Thanks.

jnothman reviewed (commented):
This was the start of a review...

thechargedneutron commented on sklearn/multioutput.py:
```diff
@@ -539,6 +474,73 @@ def predict(self, X):
         return Y_pred
 
+
+class ClassifierChain(BaseChain):
```
I did not understand which part is to be replaced with ClassifierMixin, suggest me.

jnothman replied:
It needs ClassifierMixin as an additional base class

thechargedneutron replied:
`class ClassifierChain(BaseChain, ClassifierMixin): ` 
Like this, right?

TomDLT replied:
yes

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -487,3 +487,91 @@ def test_classifier_chain_vs_independent_models():
+    Y_pred = regressor_chain.predict(X)
+
+    assert_equal(Y_pred_cv.shape, Y.shape)
+    assert_greater(jaccard_similarity_score(Y, Y_pred_cv), 0.4)
```
jaccard_similarity_score is not appropriate for regression.

thechargedneutron replied:
So, should I remove this test? or there's an alternative?

jnothman reviewed (commented):
Tests are failing, in the first instance, because you have not imported RegressorChain into test_multioutput.py

Another issue is that `classes_` should not be set in `RegressorChain` but should be in `ClassifierChain`

jnothman commented:
yes

On 2 August 2017 at 15:45, Kumar Ashutosh <notifications@github.com> wrote:

> *@thechargedneutron* commented on this pull request.
> ------------------------------
>
> In sklearn/multioutput.py
> <https://github.com/scikit-learn/scikit-learn/pull/9257#discussion_r130790144>
> :
>
> > @@ -539,6 +474,73 @@ def predict(self, X):
>
>          return Y_pred
>
> +
> +class ClassifierChain(BaseChain):
>
> class ClassifierChain(BaseChain, ClassifierMixin):
> Like this, right?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#discussion_r130790144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68xfVE9MUeeDAoFokdu0pbyeSK5Sks5sUAzxgaJpZM4OKxgm>
> .
>

jnothman commented:
Well, it's already a bit weird that we're training it on a multilabel
classification problem rather than a true regression problem, but that's no
big deal.

However, the regressor will predict continuous values, even if trained on
0s and 1s. jaccard_similarity_score only knows how to deal with 0s and 1s.
So you need a regression metric like mean_squared_error

On 2 August 2017 at 15:51, Kumar Ashutosh <notifications@github.com> wrote:

> *@thechargedneutron* commented on this pull request.
> ------------------------------
>
> In sklearn/tests/test_multioutput.py
> <https://github.com/scikit-learn/scikit-learn/pull/9257#discussion_r130790784>
> :
>
> > +
> +def test_regressor_chain_crossval_fit_and_predict():
> +    # Fit regressor chain with cross_val_predict and verify predict
> +    # performance
> +    X, Y = generate_multilabel_dataset_with_correlations()
> +    regressor_chain_cv = RegressorChain(LinearRegression(), cv=3)
> +    regressor_chain_cv.fit(X, Y)
> +
> +    regressor_chain = RegressorChain(LinearRegression())
> +    regressor_chain.fit(X, Y)
> +
> +    Y_pred_cv = regressor_chain_cv.predict(X)
> +    Y_pred = regressor_chain.predict(X)
> +
> +    assert_equal(Y_pred_cv.shape, Y.shape)
> +    assert_greater(jaccard_similarity_score(Y, Y_pred_cv), 0.4)
>
> So, should I remove this test? or there's an alternative?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#discussion_r130790784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61S5CnRvowxJmFEMbB4CStQ4gu1cks5sUA5qgaJpZM4OKxgm>
> .
>

thechargedneutron commented:
Yes, I have done all the changes suggested by you, also jaccard_similarity_score has been replaced by mean_squared_error. Any more suggestions before I commit the changes?

jnothman commented:
Do the tests pass on your machine?

you can't just replace jaccard_similarity_score by mean_squared_error. The
latter is better if smaller, and jaccard is better if greater.

On 2 August 2017 at 16:05, Kumar Ashutosh <notifications@github.com> wrote:

> Yes, I have done all the changes suggested by you, also
> jaccard_similarity_score has been replaced by mean_squared_error. Any more
> suggestions before I commit the changes?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-319576601>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-lgBK4vCztdvrnR3b73OC6gOlB0ks5sUBGcgaJpZM4OKxgm>
> .
>

thechargedneutron commented:
No, I did not try it at my local machine.
And I was not aware of the fact that jaccard_similarity_score if better if greater. So we can replace assert_greater by assert_less, and I think assert_equal and assert_not_equal should remain as it is. Am I right?

jnothman commented:
Yes, but the appropriate boundary may not be 0.4

On 2 August 2017 at 16:15, Kumar Ashutosh <notifications@github.com> wrote:

> No, I did not try it at my local machine.
> And I was not aware of the fact that jaccard_similarity_score if better if
> greater. So we can replace assert_greater by assert_less, and I think
> assert_equal and assert_not_equal should remain as it is. Am I right?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-319578115>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62wQihxAL4_Mx59y-ADrXc1_h8RZks5sUBPtgaJpZM4OKxgm>
> .
>

thechargedneutron commented:
Would a value of 0.1 be appropriate ? I tried finding if mean_squared_error is used anywhere else too, but it seems, its not used with a boundary.

jnothman commented:
No, I think we would expect MSE to be greater than .1. We want to check
that Y_pred_cv and Y_pred are​ both close to y, but different from each
other. I think you'll need to try out the code and see what extent of
difference you get.

thechargedneutron commented:
Okay, I will check it but I am not sure how will I know if a particular value is appropriate. Any suggestions.

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -509,17 +445,16 @@ def fit(self, X, Y):
         """Predict on the data matrix X using the ClassifierChain model.
 
-        Parameters
+         Parameters
```
remove the added white space

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -509,17 +445,16 @@ def fit(self, X, Y):
         X : {array-like, sparse matrix}, shape (n_samples, n_features)
             The input data.
-
         Returns
```
Add the empty line back

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -509,17 +445,16 @@ def fit(self, X, Y):
         -------
         Y_pred : array-like, shape (n_samples, n_classes)
-            The predicted values.
+        The predicted values.
```
Add the 4 spaces back

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -509,17 +445,16 @@ def fit(self, X, Y):
+        The predicted values.
 
-        """
+    """
```
Add the 4 spaces back

thechargedneutron committed:
Changes made

thechargedneutron commented:
As mentioned above, the test fails due to the presence of classes_ in BaseChain which is not used by RegressorChain, so would it be good to write the fit() function in both RegressorChain and ClassifierChain or there's other better way to achieve this? I am thinking of overriding but I am not sure

thechargedneutron committed:
RegressorChain modified

thechargedneutron committed:
Tests updated

thechargedneutron commented:
In RegressorChain class, the `fit()` function is overridden so as to avoid the error due to class_ variable.
Also, in mean_squared_error function, it was observed that the mean squared error in the default case turns out to be 0.211 . So, I set the assert_less parameter to 0.25. 
Please let me know if the value 0.25 is not appropriate.
![scikit](https://user-images.githubusercontent.com/25536866/28874495-afc3b756-77af-11e7-8026-9a2885f2ac15.png)

thechargedneutron committed:
adds Missing import

thechargedneutron commented:
The CI Tests are failing even after following the instructions above. Can anyone point out the reason for failure of tests.

jnothman commented:
The test failures are as follows, as you would have found out if you ran the tests locally, or looked at the travis logs:

```

======================================================================
ERROR: sklearn.tests.test_multioutput.test_regressor_chain_fit_and_predict_with_linear_regression
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/tests/test_multioutput.py", line 502, in test_regressor_chain_fit_and_predict_with_linear_regression
    Y_prob = regressor_chain.predict_proba(X)
AttributeError: 'RegressorChain' object has no attribute 'predict_proba'
======================================================================
FAIL: sklearn.tests.test_multioutput.test_regressor_chain_fit_and_predict_with_sparse_data
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/tests/test_multioutput.py", line 523, in test_regressor_chain_fit_and_predict_with_sparse_data
    assert_array_equal(Y_pred_sparse, Y_pred_dense)
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 718, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/numpy/testing/utils.py", line 644, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal
(mismatch 100.0%)
 x: array([[ 0.2006552 ,  0.53530139,  0.5407241 ,  0.5312952 ],
       [ 0.12791142,  0.62416716,  0.42241554,  0.24160852],
       [ 0.48269222,  0.6460722 ,  0.69942942,  0.75051546],...
 y: array([[ 0.188979  ,  0.52321891,  0.54908264,  0.50894188],
       [ 0.08790479,  0.58083609,  0.4226178 ,  0.20230125],
       [ 0.50733838,  0.63552359,  0.70484436,  0.76675438],...
>>  raise AssertionError('\nArrays are not equal\n\n(mismatch 100.0%)\n x: array([[ 0.2006552 ,  0.53530139,  0.5407241 ,  0.5312952 ],\n       [ 0.12791142,  0.62416716,  0.42241554,  0.24160852],\n       [ 0.48269222,  0.6460722 ,  0.69942942,  0.75051546],...\n y: array([[ 0.188979  ,  0.52321891,  0.54908264,  0.50894188],\n       [ 0.08790479,  0.58083609,  0.4226178 ,  0.20230125],\n       [ 0.50733838,  0.63552359,  0.70484436,  0.76675438],...')

```

thechargedneutron committed:
Updated test_multioutput.py

thechargedneutron commented:
Sorry for not going through the log. 
The function `test_regressor_chain_random_order` in `test_multioutput.py` is working fine in my local machine, but not in the build process. Is the function essential in RegressorChain? I am not sure what do I need to change in the function for the `Y_pred_random` and `Y_pred_fixed` to be same

jnothman commented:
yes, that test should work in the regression case. I'm not sure what's
going on there.


you also have flake8 failures.

On 3 Aug 2017 6:06 pm, "Kumar Ashutosh" <notifications@github.com> wrote:

> Sorry for not going through the log.
> The function test_regressor_chain_random_order in test_multioutput.py is
> working fine in my local machine, but not in the build process. Is the
> function essential in RegressorChain? I am not sure what do I need to
> change in the function for the Y_pred_random and Y_pred_fixed to be same
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-319898976>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69CN-l3NR5vF3YG9prTx5AGild8oks5sUX9_gaJpZM4OKxgm>
> .
>

jnothman committed:
Compare coef rather than predictions

jnothman commented:
I've decided to make those tests more strict, and see whether we get any clearer indications. I've pushed to your branch.

thechargedneutron committed:
Flake8 and pep8 errors removed

thechargedneutron committed:
Updated multioutput.py

thechargedneutron commented:
The error states that the array are not almost equal to 6 decimal places. We can see values like `1.18e-02` and `1.17e-02` , thus failing the tests.

![sc](https://user-images.githubusercontent.com/25536866/28921347-3ade729a-7873-11e7-84c3-ce92e2773ba1.png)

thechargedneutron commented:
Since neither predictions nor coefficients are almost equal, what should we try to do? I cannot think of a technique which must pass the conditions. Thanks.

thechargedneutron commented:
I need some help in changing `test_regressor_chain_random_order` so as to make the prediction/coefficients in fixed order and random order to be same. I tried but due to my very limited knowledge of the subject, I could not find the problem with it.

jnothman commented:
Something strange seems to be happening here on the most recent version of the dependencies...

thechargedneutron commented:
Okay, kindly ping me when it gets sorted out. Meanwhile, I am working on another issue.

jnothman commented:
Okay. I've been failing to reproduce the error, and trying to run Travis on
a local docker, but I'm at a conference this week, and large dockers are
slow to grab.

On 7 August 2017 at 06:40, Kumar Ashutosh <notifications@github.com> wrote:

> Okay, kindly ping me when it gets sorted out. Meanwhile, I am working on
> another issue.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-320531245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60bbhn6zQRqWVsNBcCBOSYonM0KZks5sViS0gaJpZM4OKxgm>
> .
>

thechargedneutron committed:
Resolved merge conflict

thechargedneutron commented:
@jnothman  Can anything be done on this PR? Any improvement in `test_regressor_chain_random_order` that you would like to be implemented so as to serve the purpose?

jnothman commented:
That's a very good question. It's sitting on my todo pile, but given the obscurity of the failure I've not yet set aside the time to open a travis in docker to try reproduce and debug the error. Are you up to doing so yourself?

thechargedneutron commented:
I am not aware of what actually needs to be done. Can you help? I went through the error log and found out the arrays are not matching upto 6 decimal places. Suggestions on how I can go about this?

jnothman commented:
Well, having been unable to reproduce the issue locally, I was going to use a Travis instance on docker to just get in there and trace back from the test failure.

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
subdtype change

thechargedneutron committed:
changes reverted

lesteve commented:
> Well, having been unable to reproduce the issue locally, I was going to use a Travis instance on docker to just get in there and trace back from the test failure.

I can not reproduce the failure either. The failure pattern does not really make sense either because the nose build fails but not the pytest one (it has exactly the same versions of numpy, scipy, etc ...) Maybe push an empty commit to see whether it was a temporary glitch.

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
removed text editor raises dissimilarities

thechargedneutron committed:
Unnecessary changes reverted

thechargedneutron committed:
empty

thechargedneutron commented:
@lesteve This is strange. Passes the Travis CI build in commit 515c2f5 but failed thereafter.

thechargedneutron committed:
empty commit 2

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron commented:
@lesteve @jnothman Test passed in 3rd attempt. Kindly review if you think this can be merged despite this strange behaviour?

jnothman commented:
Please change WIP -> MRG in the first place

thechargedneutron commented:
@jnothman Done!!

TomDLT commented on doc/modules/multiclass.rst:
```diff
@@ -425,3 +425,10 @@ averaged together.
+================
+
+Regressor chains (see :class:`RegressorChain`) are a way of combining a
+number of regressions into a single multi-label model that is capable
```
multi-label -> multi-target

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -539,6 +476,73 @@ def predict(self, X):
         return Y_pred
 
+
+class ClassifierChain(BaseChain, ClassifierMixin):
```
Why did you remove the `MetaEstimatorMixin`?

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -603,3 +607,125 @@ def decision_function(self, X):
         return Y_decision
+
+
+class RegressorChain(BaseChain, RegressorMixin):
```
You'll need a `MetaEstimatorMixin`, event though it's just a tag at the moment.

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -603,3 +607,125 @@ def decision_function(self, X):
+
+    """
+
+    def fit(self, X, Y):
```
This method is almost **entirely** duplicated in `BaseChain`, the only difference is `self.classes_` that concerns only `ClassifierChain`.

I suggest to remove the `self.classes_` from `BaseChain.fit`, in order to have in `BaseChain` the shared code.
Then I would remove `fit` from `RegressorChain`, and would have the following in `ClassifierChain`:

```py
    def fit(self, X, Y):
       """Fit the model to data matrix X and targets Y.
         Parameters
        ----------
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
            The input data.
        Y : array-like, shape (n_samples, n_classes)
            The target values.
         Returns
        -------
        self : object
            Returns self.
        """
        super(ClassifierChain, self).fit(X, Y)
        self.classes_ = []
        for chain_idx, estimator in enumerate(self.estimators_):
            self.classes_.append(estimator.classes_)
        return self
```

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -368,71 +369,7 @@ def score(self, X, y):
-    Chains for Multi-label Classification", 2009.
-
-    """
+class BaseChain(BaseEstimator):
```
We could change the name to `_BaseChain`, since this class is not supposed to be called directly

TomDLT commented on sklearn/tests/test_multioutput.py:
```diff
@@ -13,19 +13,20 @@
 from sklearn.externals.joblib import cpu_count
 from sklearn.linear_model import Lasso
-from sklearn.linear_model import LogisticRegression
+from sklearn.linear_model import LogisticRegression, LinearRegression
```
I think we tend to prefer having a single import by line, for merging conflict anticipation

thechargedneutron committed:
revised fit method in regressorchain and classifierchain

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
Basechain changed to _BaseChain

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -539,6 +473,94 @@ def predict(self, X):
+    """
+
+    def fit(self, X, Y):
+            """Fit the model to data matrix X and targets Y.
```
You can remove one level of indentation here...

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -539,6 +473,94 @@ def predict(self, X):
+            Returns
+            -------
+            self : object
+            Returns self.
```
... but here you lack one level of indentation for `Returns self`

thechargedneutron committed:
indentation corrected

thechargedneutron commented:
@TomDLT That was a very silly mistake. Sorry for that. Kindly review this commit.

TomDLT commented:
Don't worry ! You still lack one level of indentation for `Returns self` though :smile:

thechargedneutron commented:
Which line? :( I am not able to figure it out

On 20-Sep-2017 7:54 PM, "Tom Dupré la Tour" <notifications@github.com>
wrote:

> Don't worry ! You still lack one level of indentation for Returns self
> though 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-330867325>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYWpYslBSundFVA2mby65WharMl-RXUJks5skSAWgaJpZM4OKxgm>
> .
>

TomDLT commented on sklearn/multioutput.py:
```diff
@@ -539,6 +473,94 @@ def predict(self, X):
+        Returns
+        -------
+        self : object
+        Returns self.
```
Missing indentation here.
Check the other fit docstring if it's not clear

thechargedneutron committed:
Indentation corrected

thechargedneutron commented:
@TomDLT  Corrected. :)

TomDLT commented:
The documentation and the tests could probably be improved, but otherwise LGTM.
This will also need an entry in `doc/whats_new.rst`

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
Merge branch 'master' of https://github.com/thechargedneutron/scikit-learn

thechargedneutron committed:
Added whats new entry

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron commented:
@TomDLT @jnothman @lesteve  Kindly review. An entry in whats new is also added.

TomDLT commented on doc/whats_new/v0.20.rst:
```diff
@@ -40,6 +40,9 @@ Classifiers and regressors
   By :user:`Michael A. Alcorn <airalcorn2>`.
 
+- Added :class:`multioutput.RegressorChain` for multi-target
+  regression. By `Kumar Ashutosh <thechargedneutron>`_.
```
You forgot the `:user:` maker.

Can you also add the PR number: 
```
:issue:`9257`
```

thechargedneutron committed:
modified whats_new

thechargedneutron committed:
issue number

thechargedneutron committed:
removed trailin _

thechargedneutron commented:
@TomDLT Is this fine?

thechargedneutron commented:
@lesteve  @jnothman One of you kindly review this. Since this is a very old PR, it is very unlikely that any mentor will come across this directly. ;)

jnothman commented:
you can trust that it is explicitly in my pile. it is also low priority,
unfortunately. patience, please.

jnothman commented:
Regarding tests, to the extent that it is reasonable, I would rather have a test that loops over ClassifierChain and RegressorChain than have duplicated code. Are you able to try refactoring the tests?

One of the places they will differ is that I think one of your tests should assert `not hasattr(regressor_chain, 'classes_')` for the regressor chain, and the opposite for classifier chain.

jnothman reviewed (commented):
Ooh these comments seem to have been waiting for me to confirm them for some time...

thechargedneutron committed:
changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
changes added

jnothman commented:
Thanks for your quick work. I don't think you've addressed [this](https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-332484619) or [this](https://github.com/scikit-learn/scikit-learn/pull/9257#discussion_r141308331).

thechargedneutron commented on sklearn/utils/testing.py:
```diff
@@ -526,7 +526,8 @@ def uninstall_mldata_mock():
                    "OutputCodeClassifier", "OneVsRestClassifier",
-                   "RFE", "RFECV", "BaseEnsemble", "ClassifierChain"]
+                   "RFE", "RFECV", "BaseEnsemble", "ClassifierChain",
+                   "RegressorChain", "_BaseChain"]
```
@jnothman  I tried removing "_BaseChain" from here but tests failed, so restored it.

TomDLT replied:
> Use ABCMeta as in other base classes to avoid it being tested.

(https://github.com/scikit-learn/scikit-learn/pull/9257#issuecomment-347076962)

TomDLT replied:
Now you should be able to remove `_BaseChain` without test failure

thechargedneutron commented:
@jnothman  I am not able to find a way to use the same function for testing both classifier_chain and regressor_chain. There are always small differences in the test functions of these two. :(

jnothman commented:
Use `ABCMeta` as in other base classes to avoid it being tested.

thechargedneutron committed:
test_base_chain_crossval_fit_and_predict_added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
tests modified

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
space added

thechargedneutron committed:
pep8 failures removed

thechargedneutron committed:
spelling corrected

thechargedneutron committed:
order changed

thechargedneutron committed:
order changed

thechargedneutron commented:
@jnothman  Is this fine? I have looped through classifierchain and regressorchain. Please review this once.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,89 @@ def test_classifier_chain_vs_independent_models():
+              ClassifierChain(LogisticRegression())]
+    [chain.fit(X, Y) for chain in chains]
+
+    for i in range(0, 2):
```
```
for chain in chains:
    chain.fit(X, Y)
```

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,89 @@ def test_classifier_chain_vs_independent_models():
+                         RegressorChain(LinearRegression(),
+                                        order='random', random_state=42)]
+    [chain_random.fit(X, Y) for chain_random in base_chain_random]
+    base_chain_fixed = [ClassifierChain(LogisticRegression(),
```
this would be neater to read if you instead did:

```
base_chain_fixed = clone(base_chain_random).set_params(order=base_chain_random.order_)
```

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,89 @@ def test_classifier_chain_vs_independent_models():
+def test_base_chain_random_order():
+    # Fit base chain with random order
+    X, Y = generate_multilabel_dataset_with_correlations()
+    base_chain_random = [ClassifierChain(LogisticRegression(),
```
I think it would be neater if you made the entire test a loop over these rather than creating a list.

Even better, use `pytest.mark.parametrize` to run the test over each estimator. Note that you won't yet find examples of this in the codebase as we don't yet use it. But you'll find plenty of examples of looping over estimators without indexing `[i]` all the time...

thechargedneutron replied:
I looked for the usage of pytest.mark.parametrize but could not use it here since the test runs on typing `pytest` in compiler. But I could not figure out how its going to be used here. 

Reference: https://docs.pytest.org/en/latest/parametrize.html

jnothman replied:
Never mind parametrize. All I mean is:
```
for chain in [ClassifierChain(LogisticRegression()), RegressorChain(LinearRegression())]:
    ... fit chain or whatever
    chain_with_different_params = clone(chain).set_params(foo=bar)
```

thechargedneutron committed:
changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
pep8 corrected

thechargedneutron commented:
@jnothman  I guess the `clone` thing is not working because of it being a list. I agree the fact that this function could be made much better, particularly using `pytest` module, but I am not very clear with it.

thechargedneutron committed:
test_base_chain_random_order corrected

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
pep8 corrected

thechargedneutron commented:
@jnothman  All the suggested changes added. Kindly review.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+    Y_binary = (Y_prob >= .5)
+    assert_array_equal(Y_binary, Y_pred)
+
+    assert isinstance(chains[0], RegressorMixin)
```
I don't find this necessary

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+    for chain in chains:
+        chain.fit(X, Y)
+
+    for i in range(0, 2):
```
You're still using this very strange idiom. Why isn't this part of the previous loop?

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+    base_chain = [ClassifierChain(LogisticRegression(), cv=3),
+                  RegressorChain(LinearRegression(), cv=3)]
+    [chain.fit(X_sparse, Y) for chain in base_chain]
+    for i in range(0, 2):
```
Again why is this not for chain in list? Why use a range? Why use a list comprehension on the previous line when you don't use the resultant list?

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+def test_base_chain_random_order():
+    # Fit base chain with random order
+    X, Y = generate_multilabel_dataset_with_correlations()
+    for chain in [ClassifierChain(LogisticRegression()),
```
Yes, this is much better idiom

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+        chain_random.fit(X, Y)
+        chain_fixed = clone(chain).set_params(order=chain_random.order_)
+        chain_fixed.fit(X, Y)
+        assert_not_equal(list(chain_random.order), list(range(4)))
```
With pytest, we should prefer using the assert statement and comparison operators !=, ==, etc rather than these functions.

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+
+    for i in range(0, 2):
+        assert_equal(Y_pred_cv[i].shape, Y_pred[i].shape)
+    assert_greater(jaccard_similarity_score(Y, Y_pred_cv[0]), 0.4)
```
Again why should prefer bare assert

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,85 @@ def test_classifier_chain_vs_independent_models():
+            assert_array_almost_equal(est1.coef_, est2.coef_)
+
+
+def test_base_chain_crossval_fit_and_predict():
```
Given the different measurement of correctness for regressor and classifier here, it may have made sense to keep them separate. On the other hand, if you loop (or ideally parametrize) over tuples including the chain and a function to test if the similarity is good enough, this could be written in a similar idiom to what I suggest above.

thechargedneutron committed:
iterative loops removed

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

jnothman commented on sklearn/tests/test_multioutput.py:
```diff
@@ -491,3 +421,80 @@ def test_classifier_chain_vs_independent_models():
+    # Fit base chain with sparse data cross_val_predict
+    X, Y = generate_multilabel_dataset_with_correlations()
+    X_sparse = sp.csr_matrix(X)
+    base_chain = [ClassifierChain(LogisticRegression(), cv=3),
```
call this chains for consistency with above, or don't bother with the intermediate variable for consistency with below.

thechargedneutron committed:
changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

jnothman committed:
STY more succinct test (#3)

jnothman commented on sklearn/multioutput.py:
```diff
@@ -539,6 +473,94 @@ def predict(self, X):
+    Each model makes a prediction in the order specified by the chain using
+    all of the available features provided to the model plus the predictions
+    of models that are earlier in the chain.
+
```
This should say "Read more in the User Guide"

jnothman commented on sklearn/multioutput.py:
```diff
@@ -598,3 +620,61 @@ def decision_function(self, X):
+    Each model makes a prediction in the order specified by the chain using
+    all of the available features provided to the model plus the predictions
+    of models that are earlier in the chain.
+
```
This should say "Read more in the User Guide"

thechargedneutron replied:
Followed the same procedure as already there in the codebase. 
`Read more in the :ref:`User Guide <regressorchain>`.` I am not sure whether this is what you intended as the hyper-link currently may not exist.

TomDLT replied:
You need to add `.. regressorchain:` before 
```
Regressor chain
===============
```
in `doc/modules/multiclass.rst`

Same for classifier chain

jnothman reviewed (commented):
Otherwise LGTM

thechargedneutron committed:
Read more in user guide added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron committed:
doc hyperlink added

thechargedneutron committed:
metaclass added

thechargedneutron commented:
Only doc changes and test fails!! The same problem of one test failing. Let me try with an empty commit.

thechargedneutron committed:
empty commit

jnothman commented:
I agree that [test failure](https://travis-ci.org/scikit-learn/scikit-learn/jobs/313150490) is strange:
```

=================================== FAILURES ===================================
_________________________ test_base_chain_random_order _________________________
    def test_base_chain_random_order():
        # Fit base chain with random order
        X, Y = generate_multilabel_dataset_with_correlations()
        for chain in [ClassifierChain(LogisticRegression()),
                      RegressorChain(LinearRegression())]:
            chain_random = clone(chain).set_params(order='random', random_state=42)
            chain_random.fit(X, Y)
            chain_fixed = clone(chain).set_params(order=chain_random.order_)
            chain_fixed.fit(X, Y)
            assert_not_equal(list(chain_random.order), list(range(4)))
            assert_equal(len(chain_random.order_), 4)
            assert_equal(len(set(chain_random.order_)), 4)
            # Randomly ordered chain should behave identically to a fixed order
            # chain with the same order.
            for est1, est2 in zip(chain_random.estimators_,
                                  chain_fixed.estimators_):
>               assert_array_almost_equal(est1.coef_, est2.coef_)
E               AssertionError: 
E               Arrays are not almost equal to 6 decimals
E               
E               (mismatch 97.0%)
E                x: array([  7.186580e-03,  -9.998066e-03,   3.580353e-03,  -1.627024e-03,
E                       -3.538778e-03,  -2.532519e-02,   1.074851e-03,   3.214291e-02,
E                        1.339362e+11,  -1.486778e-03,   1.718521e-03,  -4.677074e+11,...
E                y: array([  7.186547e-03,  -1.018109e-02,   3.780135e-03,  -2.089836e-03,
E                       -3.220238e-03,  -2.520867e-02,   1.489337e-03,   3.183672e-02,
E                       -8.141693e+10,  -1.886845e-03,   1.585007e-03,  -5.611828e+11,...
X          = array([[ 0.83122917, -1.19111403,  1.94512639, ..., -1.04616065,
         0.19652082, -0.71217646],
       [-1.7883219...266, -1.00619016],
       [-1.08628488, -1.20595377, -1.19448098, ...,  0.88972995,
        -1.08459736, -1.89365993]])
Y          = array([[0, 0, 1, 1],
       [0, 1, 1, 0],
       [1, 0, 1, 1],
       ..., 
       [1, 1, 1, 0],
       [0, 0, 1, 0],
       [0, 1, 0, 0]])
chain      = RegressorChain(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False),
        cv=None, order=None, random_state=None)
chain_fixed = RegressorChain(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False),
        cv=None, order=array([1, 3, 0, 2]), random_state=None)
chain_random = RegressorChain(base_estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False),
        cv=None, order='random', random_state=42)
est1       = LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)
est2       = LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)
/home/travis/build/scikit-learn/scikit-learn/sklearn/tests/test_multioutput.py:471: AssertionError
====== 1 failed, 8351 passed, 26 skipped, 4860 warnings in 544.24 seconds ======
```

It only occurs on one platform, only sometimes, and only in the regression case. Which would suggest it is an indeterminacy in `LinearRegression` which is quite unbelievable. @tomdlt any ideas?

I doubt it will change anything, but there should be a check there that `chain_fixed.order_` and `chain_random.order_` are identical. Could you add that?

thechargedneutron committed:
added check

thechargedneutron commented:
@jnothman  Yeah, added that. Works fine locally.

TomDLT commented:
This error is weird.
I set up a conda environment with the same versions as Travis' 3rd job (python 3.6.2) but could not reproduce locally (Linux-4.4.0-98-generic-x86_64-with-debian-stretch-sid).

> Which would suggest it is an indeterminacy in LinearRegression which is quite unbelievable.

Indeed !

thechargedneutron committed:
META_ESTIMATORS removed

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn

thechargedneutron commented:
@TomDLT  Tests failing. I guess some more change needs to be done apart from adding ABCMeta. What needs to be done in addition to it. I don't have much experience in this.

TomDLT committed:
FIX remove _BaseChain from all_estimators by adding an abstract method

TomDLT committed:
investigate failure on Travis 3.6.1

TomDLT commented:
I removed `_BaseChain` from `all_estimators` by adding an abstract method, which fixed the `estimator_check` error.
___
About Travis' failure on python 3.6.1, it seems that `LinearRegression` is not deterministic, as shown by the following test:
```py
def test_investigate_linear_regression_indeterminacy():
    # Is LinearRegression deterministic?
    X, Y = generate_multilabel_dataset_with_correlations()
    y = Y[:, 1]
    ref = LinearRegression().fit(X, y).coef_
    for i in range(1000):
        coef = LinearRegression().fit(X, y).coef_
        assert_array_equal(ref, coef, 'iter %d' % i)
```
which fails **at iteration 18** in [this Travis job](https://travis-ci.org/scikit-learn/scikit-learn/jobs/314812527#L2797).

This failure is not related to this PR, but if we want to merge it, I would rather not merge `test_base_chain_random_order` as it is (and remove `test_investigate_linear_regression_indeterminacy`).

TomDLT committed:
investigate failure on scipy.linalg.lstsq

TomDLT committed:
FIX give float64 array to lstsq

TomDLT committed:
Remove investigation test and change LinearRegression into Ridge

jnothman commented:
Thanks @thechargedneutron

amueller commented on doc/modules/multiclass.rst:
```diff
@@ -425,3 +427,13 @@ averaged together.
     Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank,
         "Classifier Chains for Multi-label Classification", 2009.
+
+.. regressorchain:
```
needs ``_`` in the front

thechargedneutron replied:
Should I submit a new PR with these changes?

amueller replied:
yeah, sounds good. And mention that in #10258 and check the circle CI build whether the links work. Maybe also add a "see also" from the regressor to the classifier and vice versa.
